File: examples/crater_self_shadow/fn.py
#!/usr/bin/env python

import numpy

from kalast.app import App, Hud, Simulation


def before_render(sim: Simulation, dt: float) -> None:
    it = sim.state.iteration
    if it >= 10000:
        return

    a = it * 0.005
    sim.sun.pos = [0.0, 20.0 * numpy.sin(a), 20.0 * numpy.cos(a)]


def after_render(sim: Simulation, dt: float) -> None:
    it = sim.state.iteration
    if it >= 10000:
        return

    illum = sim.facet_illumination(0)
    lit = float((illum > 0).mean()) if illum is not None else 0.0
    sim.huds[0].text = f"lit {lit * 100:.1f} %"


app = App()
app.simulation.config.axes.style = "gizmo"
app.simulation.config.light.cube_show = True
app.simulation.config.shading.render_back_face = True
app.simulation.config.shadows.access_shadow_map = True
app.simulation.config.wireframe.mode = 2
app.simulation.config.wireframe.color = [0.05, 0.05, 0.05, 1.0]
# app.simulation.config.shadows.resolution = 8192
# app.simulation.config.shadows.pcf = 2
app.simulation.huds = [Hud("", size=16)]
app.simulation.camera.pos = [1.5, 2.0, 1.5]
app.simulation.camera.look_anchor()
app.simulation.load_mesh(path="res/plane_crater_1024-5000_h=0.437.obj")

# See main.py to call kalast in main while loop iteration by iteration.
app.before_render = before_render
app.after_render = after_render
app.start()

File: examples/crater_self_shadow/fn.rs
use std::cell::RefCell;
use std::rc::Rc;

use kalast::app::App;
use kalast::app::config::Hud;
use kalast::app::simulation::Simulation;
use kalast::{Float, Mat4, Vec3};

fn main() {
    let mut app = App::new();

    {
        let config = app.sim_config();
        let mut c = config.borrow_mut();
        c.axes.style = kalast::app::axes::AxesStyle::Gizmo;
        c.shading.render_back_face = true;
        c.light.cube_show = true;
        c.shadows.access_shadow_map = true;
        c.wireframe.mode = 2;
        c.wireframe.color = wgpu::Color {
            r: 0.05,
            g: 0.05,
            b: 0.05,
            a: 1.0,
        };
    }

    {
        let mut sim = app.simulation.borrow_mut();

        let mut hud = Hud::new("");
        hud.size = 16.0;
        sim.huds = vec![Rc::new(RefCell::new(hud))];

        sim.camera.pos = Vec3::new(1.5778934, 1.9384689, 1.5082116);
        sim.camera.look_anchor();

        sim.load_mesh(
            "res/plane_crater_1024-5000_h=0.437.obj",
            Mat4::IDENTITY,
            false,
        );
    }

    // Using closures for before_ and after_render but can be written in plain
    // separated functions like main.py could have been lambda functions aswell.

    app.set_before_render(|sim: &mut Simulation, _dt: Float| {
        let a = sim.state.iteration as Float * 0.005;
        sim.sun.pos = Vec3::new(0.0, 20.0 * a.sin(), 20.0 * a.cos());
    });

    app.set_after_render(|sim: &mut Simulation, _dt: Float| {
        let lit = match sim.facet_illumination(0) {
            Some(illum) if !illum.is_empty() => {
                illum.iter().filter(|&&i| i > 0.0).count() as Float / illum.len() as Float
            }
            _ => 0.0,
        };
        let it = sim.state.iteration;
        sim.huds[0].borrow_mut().text = format!("lit {:.1} %", lit * 100.0);
    });

    app.start();
}
